pkg/controller/statefulset/stateful_set_control.go

```diff
@@ -516,9 +517,20 @@ func (ssc *defaultStatefulSetControl) updateStatefulSet(
 
 	// we compute the minimum ordinal of the target sequence for a destructive update based on the strategy.
 	updateMin := 0
+	maxUnavailable := 1
 	if set.Spec.UpdateStrategy.RollingUpdate != nil {
 		updateMin = int(*set.Spec.UpdateStrategy.RollingUpdate.Partition)
+		maxUnavailable, err = intstrutil.GetValueFromIntOrPercent(intstrutil.ValueOrDefault(set.Spec.UpdateStrategy.RollingUpdate.MaxUnavailable, intstrutil.FromInt(1)), int(replicaCount), false)
```

Comment:
yeah, but that variable is still needed to track the value and initialize it if the case is RollingUpdate ==nil, in which case it wont come inside this block

Reply:
What happens if an object gets `maxUnavailable = 2`, and then the feature gate is disabled?

From this code (and the API strategy), it seems that the controller would still honor the value of 2. Is this what we want?

Reply:
if the feature gate is disabled, then the drop fields code should drop the value and hence maxUavailable will be treated as 1 which is what the implicit behavior is today without this feature. Are you suggesting this code should be under a feature flag ? My impression is that the drop fields would return the right value here

Reply:
consider a case where the feature gate is enabled and later disabled. In this case, the apiserver wouldn't drop the field.

Reply:
If its eenabled and disabled again, we should treat that as with value 1. How do i ensure that, since the stored value is 2 now . What is the general pattern for other features to handle this ?

Reply:
That's outside the scope of the api-review (which I'm providing).

I would think that querying the feature gate in the controller-manager is the way to do it, but @kow3ns for the authoritative answer.

Reply:
Essentially the controller should resort to old behavior which is exactly the same as maxUnavailable=1. Now the question is does the api machinery code have this covered(1: take a default value of a field 2: auto reset that field to the default value anytime thee feature gets disabled, whether first time or after getting enabled.) . If not covered, is that ok if handle that by putting an explicit code here saying` if feature gate disabled, maxUnavailable := 1`, wouuld that be ok ?

Reply:
@kow3ns @lavalamp looking for how to proceed here. Does api-machinery know to reset the value of a field to the default value when feature gate is disabled or went from enabled to disabled or do i need to handdle this explicitly here by saying `if feature gate is disabled set maxUnavailable := 1(default value) ?

Reply:
> Now the question is does the api machinery code have this covered

The strategy is only applied when an object is being updated. So answer is no.

Reply:
I think only generic validation for the field should be done in  case of enabled, disabled and re-enabled but say if the user has explicitly to 2 while the featuregate is disabled and after that the featuegate has been enabled, the STS should respect the value of 2.

Reply:
i have two questions here
1: how do i respect the value of 2 ? If its enabled and value is not 1, do nothing  otherwise the default(via defaulting) value of maxUnavailable will apply anyway without any additional work
2: why this would be helpful ? I think keeping it at the default value when moving from disabled to enabled will avoid surprises but i can see some people may not like this view and may want everything to just work as it was when it was enabled

Reply:
why are we still discussing this? We already decided: https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/961-maxunavailable-for-statefulset#upgradesdowngrades

tl;dr: leave the value untouched, but have the controller ignore the field.

Reply:
Untouched + ignoring is what I mentioned ^.